### PR TITLE
feat: introduce source maps for templates

### PR DESCRIPTION
**Pull request by tbosch**
Last part of https://github.com/angular/angular/issues/14013.

### CODE REVIEW

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 1]
```diff
@@ -738,3 +742,42 @@ export function flatten<T>(list: Array<T|T[]>): T[] {
...
+      return compMeta.type.reference.filePath;
+    } else {
+      return `${ngJitFolder()}/${identifierName(ngModuleType)}/${identifierName(compMeta.type)}.html`;
+    }
+  } else {
+    return templateMeta.templateUrl;
+  }
+}
+
+export function sharedStylesheetJitUrl(meta: CompileStylesheetMetadata, id: number) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/compile_metadata.ts]: add more unit tests for this.
Maybe use `lastIndexOf` to simplify...

File: modules/@angular/compiler/src/output/output_ast.ts [hunk 2]
```diff
@@ -912,33 +1001,65 @@ export class RecursiveExpressionVisitor implements StatementVisitor, ExpressionV
...
-}
-
-class _ReplaceVariableTransformer extends ExpressionTransformer {
-  constructor(private _varName: string, private _newValue: Expression) { super(); }
-  visitReadVarExpr(ast: ReadVarExpr, context: any): any {
-    return ast.name == this._varName ? this._newValue : ast;
-  }
-}
-
 export function findReadVarNames(stmts: Statement[]): Set<string> {
```
Comment (tbosch) [inline modules/@angular/compiler/src/output/output_ast.ts]: drop

File: modules/@angular/compiler/src/output/output_ast.ts [hunk 3]
```diff
@@ -912,33 +1001,65 @@ export class RecursiveExpressionVisitor implements StatementVisitor, ExpressionV
...
   const finder = new _VariableFinder();
   finder.visitAllStatements(stmts, null);
   return finder.varNames;
 }
 
-class _VariableFinder extends RecursiveExpressionVisitor {
+class _VariableFinder extends RecursiveAstVisitor {
```
Comment (tbosch) [inline modules/@angular/compiler/src/output/output_ast.ts]: drop

File: modules/@angular/compiler/test/aot/compiler_spec.ts [hunk 4]
```diff
@@ -68,6 +71,154 @@ describe('compiler (unbundled Angular)', () => {
              .toBeDefined();
        })));
   });
+
+  describe('aot source mapping', () => {
+    const componentPath = '/app/app.component.ts';
+
+    let mockData: MockData;
```
Comment (tbosch) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: maybe call `rootDir`

File: modules/@angular/compiler/test/aot/compiler_spec.ts [hunk 5]
```diff
@@ -68,6 +71,154 @@ describe('compiler (unbundled Angular)', () => {
...
+          ${componentDecorator}
+        })
+        export class AppComponent {
+          someMethod() {}
+        }
+      `;
+    }
+
+    describe('inline templates', () => {
+      const templateUrl = '/app/app.component.ts';
```
Comment (tbosch) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: use componentPath

File: modules/@angular/compiler/test/aot/compiler_spec.ts [hunk 6]
```diff
@@ -68,6 +71,154 @@ describe('compiler (unbundled Angular)', () => {
...
+           expect(() => compileApp())
+               .toThrowError(new RegExp(`Template parse errors[\\s\\S]*${templateUrl}@1:7`));
+         }));
+
+      it('should create a sourceMap for the template', fakeAsync(() => {
+           const template = 'Hello World!';
+
+           appDir['app.component.ts'] = createComponentSource(templateDecorator(template));
+
+           const genFile = compileApp().find(genFile => genFile.srcFileUrl === componentPath);
```
Comment (tbosch) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: Move this into `compileApp`

File: modules/@angular/compiler/test/aot/compiler_spec.ts [hunk 7]
```diff
@@ -68,6 +71,154 @@ describe('compiler (unbundled Angular)', () => {
...
+           const sourceMap = extractSourceMap(genFile.source);
+           // the generated file contains the host view and the component view.
+           // we are only interested in the component view.
+           const sourceIndex = sourceMap.sources.indexOf(templateUrl);
+           expect(sourceMap.file).toEqual(genFile.genFileUrl);
```
Comment (tbosch) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: move one line up

File: modules/@angular/compiler/test/aot/test_util.ts [hunk 8]
```diff
@@ -169,7 +169,9 @@ export class MockCompilerHost implements ts.CompilerHost {
   private assumeExists = new Set<string>();
   private traces: string[] = [];
 
-  constructor(scriptNames: string[], private data: MockData, private angular: Map<string, string>) {
```
Comment (tbosch) [inline modules/@angular/compiler/test/aot/test_util.ts]: remove this change

File: modules/@angular/core/src/view/services.ts [hunk 9]
```diff
@@ -357,13 +357,7 @@ class DebugContext_ implements DebugContext {
     }
     return references;
   }
-  get source(): string {
```
Comment (tbosch) [inline modules/@angular/core/src/view/services.ts]: remove completely from here, and move message into `DebugNode`.

also: add message into commit message.

File: modules/@angular/core/src/view/services.ts [hunk 10]
```diff
@@ -372,6 +366,32 @@ class DebugContext_ implements DebugContext {
...
+        return console.error.bind(console, ...values);
+      } else {
+        return NOOP;
+      }
+    };
+    logViewFactory(nodeLogger);
+    if (currNodeIndex < logNodeIndex) {
+      // Somehow the logger was not called.
+      // So call it ourselves...
+      (<any>console.error)(...values);
```
Comment (tbosch) [inline modules/@angular/core/src/view/services.ts]: Log a message: `ILLEGAL STATE`...

File: modules/playground/src/bootstrap.ts [hunk 11]
```diff
@@ -10,7 +10,7 @@
 (function(global: any) {
   writeScriptTag('/all/playground/vendor/core.js');
   writeScriptTag('/all/playground/vendor/zone.js');
-  writeScriptTag('/all/playground/vendor/long-stack-trace-zone.js');
+  // writeScriptTag('/all/playground/vendor/long-stack-trace-zone.js');
```
Comment (tbosch) [inline modules/playground/src/bootstrap.ts]: Undo this.

Comment (tbosch) [inline modules/playground/src/bootstrap.ts]: Link https://github.com/angular/zone.js/issues/661 in commit mesage, that it does not work with long stack traces.

**Review by vicb — APPROVED** [review]

File: modules/@angular/compiler/src/compile_metadata.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/compile_metadata.ts]: Done

File: modules/@angular/compiler/src/output/output_ast.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/output/output_ast.ts]: Done

File: modules/@angular/compiler/src/output/output_ast.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/compiler/src/output/output_ast.ts]: Done

File: modules/@angular/compiler/test/aot/compiler_spec.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: Done

File: modules/@angular/compiler/test/aot/compiler_spec.ts (hunk 5 above)
Comment (tbosch) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: Done

File: modules/@angular/compiler/test/aot/compiler_spec.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: Done

File: modules/@angular/compiler/test/aot/compiler_spec.ts (hunk 7 above)
Comment (tbosch) [inline modules/@angular/compiler/test/aot/compiler_spec.ts]: Done

File: modules/@angular/compiler/test/aot/test_util.ts (hunk 8 above)
Comment (tbosch) [inline modules/@angular/compiler/test/aot/test_util.ts]: Done

File: modules/@angular/core/src/view/services.ts (hunk 9 above)
Comment (tbosch) [inline modules/@angular/core/src/view/services.ts]: Done

File: modules/@angular/core/src/view/services.ts (hunk 10 above)
Comment (tbosch) [inline modules/@angular/core/src/view/services.ts]: Done

File: modules/playground/src/bootstrap.ts (hunk 11 above)
Comment (tbosch) [inline modules/playground/src/bootstrap.ts]: Done

**Review by IgorMinar — APPROVED** [review]
lgtm for api changes. we don't expect any impact of this but I'll give heads up to Augury.

**Review by mhevery — APPROVED** [review]

**Review by vicb — APPROVED** [review]

**Pull request merged**